File: backend/src/agents/providers/ollama_cloud.py
from __future__ import annotations

import json
from typing import Any, AsyncGenerator, Optional

import httpx

from src.agents.providers.base import ProviderStreamDelta
from src.agents.providers.openai_compatible import OpenAICompatibleProvider
from src.schemas.providers import ProviderMetadata, ProviderModel


class OllamaCloudError(Exception):
    pass


class OllamaCloudProvider(OpenAICompatibleProvider):
    def __init__(self, metadata: ProviderMetadata) -> None:
        super().__init__(metadata)

    async def list_models(self, api_key: str, base_url: Optional[str] = None) -> list[ProviderModel]:
        endpoint = "https://ollama.com/api/tags"
        headers = self._headers(api_key)
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.get(endpoint, headers=headers)
            response.raise_for_status()
            payload = response.json()

        items = payload.get("models", [])
        models: list[ProviderModel] = []
        for item in items:
            model_id = item.get("name")
            if not model_id:
                continue
            models.append(
                ProviderModel(
                    id=model_id,
                    provider=self.metadata.id,
                    label=model_id,
                    owned_by="ollama",
                    supports_tools=True,
                )
            )
        models.sort(key=lambda model: model.label.lower())
        return models

    async def stream_chat_completion(
        self,
        *,
        api_key: str,
        model: str,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]],
        base_url: Optional[str] = None,
        temperature: float = 0.2,
    ) -> AsyncGenerator[ProviderStreamDelta, None]:
        endpoint = "https://ollama.com/api/chat"
        headers = self._headers(api_key)
        clean_messages = self._normalize_messages(messages)
        clean_tools = self._normalize_tools(tools)
        payload: dict[str, Any] = {
            "model": model,
            "messages": clean_messages,
            "stream": True,
        }
        if clean_tools:
            payload["tools"] = clean_tools

        async with httpx.AsyncClient(timeout=httpx.Timeout(90.0, connect=30.0)) as client:
            async with client.stream("POST", endpoint, headers=headers, json=payload) as response:
                if response.status_code >= 400:
                    body = await response.aread()
                    error_msg = f"Ollama Cloud error ({response.status_code})"
                    try:
                        data = json.loads(body)
                        if err := data.get("error"):
                            error_msg = f"Ollama Cloud: {err}"
                    except (json.JSONDecodeError, UnicodeDecodeError):
                        text = body.decode(errors="replace").strip()
                        if text:
                            error_msg = f"Ollama Cloud error ({response.status_code}): {text}"
                    raise OllamaCloudError(error_msg)

                async for line in response.aiter_lines():
                    if not line:
                        continue
                    try:
                        event = json.loads(line)
                    except json.JSONDecodeError:
                        continue
                    if err := event.get("error"):
                        raise OllamaCloudError(str(err))
                    yield self._delta_from_event(event)

    def _normalize_messages(self, messages: list[dict[str, Any]]) -> list[dict[str, Any]]:
        clean: list[dict[str, Any]] = []
        for msg in messages:
            entry: dict[str, Any] = {"role": msg["role"]}
            content = msg.get("content")
            if content is not None:
                entry["content"] = content

            if msg["role"] == "assistant":
                tool_calls = msg.get("tool_calls")
                if tool_calls:
                    entry["tool_calls"] = self._normalize_tool_calls(tool_calls)
            elif msg["role"] == "tool":
                entry["tool_name"] = msg.get("name", "")
            clean.append(entry)
        return clean

    def _normalize_tool_calls(self, tool_calls: list[dict[str, Any]]) -> list[dict[str, Any]]:
        normalized: list[dict[str, Any]] = []
        for tc in tool_calls:
            func = tc.get("function", {})
            args = func.get("arguments", "{}")
            if isinstance(args, str):
                try:
                    args = json.loads(args or "{}")
                except json.JSONDecodeError:
                    args = {}
            normalized.append({
                "function": {
                    "name": func.get("name", ""),
                    "arguments": args,
                }
            })
        return normalized

    def _normalize_tools(self, tools: list[dict[str, Any]]) -> list[dict[str, Any]]:
        clean: list[dict[str, Any]] = []
        for tool in tools:
            func = tool.get("function", {})
            clean.append({
                "type": tool.get("type", "function"),
                "function": {
                    "name": func.get("name", ""),
                    "description": func.get("description", ""),
                    "parameters": func.get("parameters", {}),
                },
            })
        return clean

    def _delta_from_event(self, event: dict[str, Any]) -> ProviderStreamDelta:
        message = event.get("message") or {}
        content = self._extract_text(message.get("content"))
        reasoning = self._extract_text(
            message.get("thinking") or message.get("reasoning") or message.get("reasoning_content") or ""
        )
        raw_tool_calls = message.get("tool_calls") or None
        tool_calls = self._openai_style_tool_calls(raw_tool_calls)
        done = event.get("done", False)
        finish_reason = event.get("done_reason") if done else None

        return ProviderStreamDelta(
            text=content if content else None,
            reasoning=reasoning if reasoning else None,
            tool_calls=tool_calls,
            finish_reason=finish_reason,
            raw=event,
        )

    def _openai_style_tool_calls(self, tool_calls: Any) -> Any:
        if not tool_calls:
            return None
        normalized: list[dict[str, Any]] = []
        for i, tc in enumerate(tool_calls):
            func = tc.get("function", {})
            args = func.get("arguments", {})
            if isinstance(args, dict):
                args = json.dumps(args)
            normalized.append({
                "index": i,
                "id": f"call_{i}",
                "type": "function",
                "function": {
                    "name": func.get("name", ""),
                    "arguments": args,
                },
            })
        return normalized

File: curro-ai/src/agents/providers/ollamaCloud.ts
import { OpenAICompatibleProvider } from "./base.js";
import type {
  ChatCompletionOptions,
  ProviderModel,
  StreamDelta,
  ToolCallDelta,
} from "./types.js";

/**
 * Ollama Cloud is NOT OpenAI-compatible: models come from `/api/tags` and chat streams
 * newline-delimited JSON from `/api/chat` (not SSE). We normalize both to the common
 * Provider surface so the agent loop treats it like any other provider.
 */
class OllamaCloudProvider extends OpenAICompatibleProvider {
  /** Ollama's REST root (e.g. https://ollama.com/api), derived from the OpenAI-style base URL. */
  private apiRoot(override?: string): string {
    const raw = (override || this.metadata.defaultBaseUrl).replace(/\/+$/, "");
    return raw.replace(/\/v1$/, "");
  }

  override async listModels(apiKey: string, baseUrl?: string): Promise<ProviderModel[]> {
    const response = await fetch(`${this.apiRoot(baseUrl)}/tags`, {
      method: "GET",
      headers: this.headers(apiKey),
    });
    if (!response.ok) {
      throw new Error(`${this.metadata.label} models error ${response.status}: ${await safeText(response)}`);
    }
    const payload = (await response.json()) as { models?: Array<Record<string, unknown>> };
    const items = payload.models ?? [];

    const models: ProviderModel[] = [];
    for (const item of items) {
      const id = item.name as string;
      if (!id) continue;
      models.push({
        id,
        provider: this.metadata.id,
        label: id,
        owned_by: "ollama",
        context_window: null,
      });
    }
    models.sort((a, b) => a.label.toLowerCase().localeCompare(b.label.toLowerCase()));
    return models;
  }

  override async *streamChatCompletion(options: ChatCompletionOptions): AsyncGenerator<StreamDelta, void, unknown> {
    const body: Record<string, unknown> = {
      model: options.model,
      messages: normalizeMessages(options.messages),
      stream: true,
    };
    const tools = normalizeTools(options.tools);
    if (tools.length > 0) body.tools = tools;

    const response = await fetch(`${this.apiRoot(options.baseUrl)}/chat`, {
      method: "POST",
      headers: this.headers(options.apiKey),
      body: JSON.stringify(body),
      signal: options.signal,
    });

    if (!response.ok || !response.body) {
      let message = `${this.metadata.label} error (${response.status})`;
      const text = await safeText(response);
      if (text) {
        const parsed = tryParse(text);
        if (parsed && typeof parsed.error === "string") message = `${this.metadata.label}: ${parsed.error}`;
        else message = `${this.metadata.label} error (${response.status}): ${text}`;
      }
      throw new Error(message);
    }

    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        buffer += decoder.decode(value, { stream: true });

        let newlineIndex: number;
        while ((newlineIndex = buffer.indexOf("\n")) !== -1) {
          const line = buffer.slice(0, newlineIndex).trim();
          buffer = buffer.slice(newlineIndex + 1);
          if (!line) continue;

          const event = tryParse(line);
          if (!event) continue;
          if (typeof event.error === "string") throw new Error(String(event.error));
          yield deltaFromEvent(event);
        }
      }

      const tail = buffer.trim();
      if (tail) {
        const event = tryParse(tail);
        if (event) {
          if (typeof event.error === "string") throw new Error(String(event.error));
          yield deltaFromEvent(event);
        }
      }
    } finally {
      reader.releaseLock();
    }
  }
}

function deltaFromEvent(event: Record<string, unknown>): StreamDelta {
  const message = (event.message as Record<string, unknown>) ?? {};
  const text = extractText(message.content);
  const reasoning = extractText(message.thinking ?? message.reasoning ?? message.reasoning_content ?? "");
  const toolCalls = openAiStyleToolCalls(message.tool_calls);
  const done = event.done === true;
  const finishReason = done ? ((event.done_reason as string) ?? "stop") : null;

  return {
    text: text || undefined,
    reasoning: reasoning || undefined,
    toolCalls,
    finishReason,
  };
}

/** Convert Ollama's tool-call objects into OpenAI-style streaming deltas. */
function openAiStyleToolCalls(raw: unknown): ToolCallDelta[] | undefined {
  if (!Array.isArray(raw) || raw.length === 0) return undefined;
  return raw.map((tc, index) => {
    const func = (tc?.function as Record<string, unknown>) ?? {};
    const args = func.arguments;
    return {
      index,
      id: `call_${index}`,
      type: "function",
      function: {
        name: (func.name as string) ?? "",
        arguments: typeof args === "string" ? args : JSON.stringify(args ?? {}),
      },
    };
  });
}

/** Reshape OpenAI-style messages into the structure Ollama's /api/chat expects. */
function normalizeMessages(messages: Array<Record<string, unknown>>): Array<Record<string, unknown>> {
  return messages.map((msg) => {
    const role = msg.role as string;
    const entry: Record<string, unknown> = { role };
    if (msg.content !== undefined && msg.content !== null) entry.content = msg.content;

    if (role === "assistant" && Array.isArray(msg.tool_calls)) {
      entry.tool_calls = (msg.tool_calls as Array<Record<string, unknown>>).map((tc) => {
        const func = (tc.function as Record<string, unknown>) ?? {};
        let args: unknown = func.arguments ?? {};
        if (typeof args === "string") {
          args = tryParse(args) ?? {};
        }
        return { function: { name: (func.name as string) ?? "", arguments: args } };
      });
    } else if (role === "tool") {
      entry.tool_name = (msg.name as string) ?? "";
    }
    return entry;
  });
}

/** Strip OpenAI tool schemas down to what Ollama accepts. */
function normalizeTools(tools: unknown[]): Array<Record<string, unknown>> {
  if (!Array.isArray(tools)) return [];
  return tools.map((tool) => {
    const t = (tool as Record<string, unknown>) ?? {};
    const func = (t.function as Record<string, unknown>) ?? {};
    return {
      type: (t.type as string) ?? "function",
      function: {
        name: (func.name as string) ?? "",
        description: (func.description as string) ?? "",
        parameters: (func.parameters as Record<string, unknown>) ?? {},
      },
    };
  });
}

function extractText(value: unknown): string {
  if (value == null) return "";
  if (typeof value === "string") return value;
  if (Array.isArray(value)) {
    return value
      .map((item) => {
        if (item && typeof item === "object") {
          const record = item as Record<string, unknown>;
          return String(record.text ?? record.content ?? "");
        }
        return String(item);
      })
      .join("");
  }
  return String(value);
}

function tryParse(text: string): Record<string, unknown> | null {
  try {
    const parsed = JSON.parse(text);
    return parsed && typeof parsed === "object" ? (parsed as Record<string, unknown>) : null;
  } catch {
    return null;
  }
}

async function safeText(response: Response): Promise<string> {
  try {
    const text = await response.text();
    return text.slice(0, 500);
  } catch {
    return "";
  }
}

export const ollamaCloudProvider = new OllamaCloudProvider({
  id: "ollama_cloud",
  label: "Ollama Cloud",
  defaultBaseUrl: "https://ollama.com/api/v1",
});
